numpyro/infer/hmc_gibbs.py

```diff
@@ -149,3 +153,178 @@ def potential_fn(z_gibbs, z_hmc):
+        ...     numpyro.sample("x", dist.Normal(locs[c], 0.5))
+        ...
+        >>> probs = jnp.array([0.15, 0.3, 0.3, 0.25])
+        >>> locs = jnp.array([-2, 0, 2, 4])
+        >>> gibbs_fn = discrete_gibbs_fn(model, probs, locs)
+        >>> kernel = HMCGibbs(NUTS(model), gibbs_fn, gibbs_sites=["c"])
+        >>> mcmc = MCMC(kernel, 1000, 100000, progress_bar=False)
+        >>> mcmc.run(random.PRNGKey(0), probs, locs)
+        >>> samples = mcmc.get_samples()["x"]
+        >>> assert abs(jnp.mean(samples) - 1.3) < 0.1
+
+    """
+    # NB: all of the information such as `model`, `model_args`, `model_kwargs`
+    # can be accessed from HMCGibbs.sample but we require them here to
+    # simplify the api of `gibbs_fn`
+    prototype_trace = trace(seed(model, rng_seed=0)).get_trace(*model_args, **model_kwargs)
+    support_sizes = {
+        name: jnp.broadcast_to(site["fn"].enumerate_support(False).shape[0], jnp.shape(site["value"]))
+        for name, site in prototype_trace.items()
+        if site["type"] == "sample" and site["fn"].has_enumerate_support and not site["is_observed"]
+    }
+
+    def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
+        # convert to unconstrained values
+        z_hmc = {k: biject_to(prototype_trace[k]["fn"].support).inv(v)
+                 for k, v in hmc_sites.items() if k in prototype_trace}
+        enum = any(site["type"] == "sample"
+                   and not site["is_observed"]
+                   and site["fn"].has_enumerate_support
+                   and name not in gibbs_sites
+                   for name, site in prototype_trace.items())
+
+        def potential_fn(z_discrete):
+            model_kwargs["_gibbs_sites"] = z_discrete
+            return potential_energy(_wrap_model(model), model_args, model_kwargs, z_hmc, enum=enum)
+
+        # get support_sizes of gibbs_sites
+        support_sizes_flat, _ = ravel_pytree({k: support_sizes[k] for k in gibbs_sites})
+        num_discretes = support_sizes_flat.shape[0]
+
+        rng_key, rng_permute = random.split(rng_key)
+        idxs = random.permutation(rng_key, jnp.arange(num_discretes))
+
+        def body_fn(i, val):
+            idx, support_size = idxs[i], support_sizes_flat[i]
+            return _discrete_step(*val, potential_fn=potential_fn, idx=idx, support_size=support_size)
+
+        init_val = (rng_key, gibbs_sites, potential_fn(gibbs_sites))
+        _, gibbs_sites, _ = fori_loop(0, num_discretes, body_fn, init_val)
+        return gibbs_sites
+
+    return gibbs_fn
+
+
+def subsample_gibbs_fn(model, *model_args, **model_kwargs):
+    """
+    [EXPERIMENTAL INTERFACE]
+
+    Returns a gibbs_fn to be used in :class:`HMCGibbs`, which works for subsampling
+    statements using :class:`~numpyro.plate` primitive. This implements the Algorithm 1
+    of reference [1] but uses a naive estimation (without control variates) of log likelihood,
+    hence might incur a high bias (reference [2]).
```

Comment:
bias-> variance?
isn't this asymptotically exact?

Reply:
I am not sure if this is asymptotically exact. [This reference](https://arxiv.org/abs/1404.4178) has a proof for convergences when control variates are used.